File: frontend/src/mocks/fixtures/postInferenceBins.ts
import { publicDeviceId, type SensorId } from '../../contracts/common'
import type { PostInferenceBin } from '../../contracts/postInferenceBins'

function bin(
  ordinal: number,
  start: string,
  end: string,
  isAlert: boolean,
): Readonly<PostInferenceBin> {
  return Object.freeze({
    segment_id: 0,
    bin_ordinal: ordinal,
    start_score_ts: start,
    end_score_ts: end,
    scored_timestamp_count: 51,
    is_alert: isAlert,
    candidate_alert_count: isAlert ? 3 : 0,
    first_alert_ts: isAlert ? start : null,
    last_alert_ts: isAlert ? end : null,
    peak_score: isAlert ? 1.4 : 0.6,
    latest_score: isAlert ? 1.1 : 0.5,
    threshold: 1,
    schema_version: 'post-inference-bins-v1',
  })
}

export const normalPostInferenceBinsBySensor = Object.freeze({
  [publicDeviceId]: Object.freeze([
    bin(0, '2026-05-31T23:47:30', '2026-05-31T23:49:30', false),
    bin(1, '2026-05-31T23:50:00', '2026-05-31T23:52:00', false),
    bin(2, '2026-05-31T23:52:30', '2026-05-31T23:54:30', true),
  ]),
} satisfies Record<SensorId, readonly PostInferenceBin[]>)

export const alertPostInferenceBinsBySensor = Object.freeze({
  [publicDeviceId]: Object.freeze([
    bin(0, '2026-05-31T23:47:30', '2026-05-31T23:49:30', true),
    bin(1, '2026-05-31T23:50:00', '2026-05-31T23:52:00', true),
    bin(2, '2026-05-31T23:52:30', '2026-05-31T23:54:30', true),
  ]),
} satisfies Record<SensorId, readonly PostInferenceBin[]>)

export const normalPostInferenceBins = normalPostInferenceBinsBySensor[publicDeviceId]
export const alertPostInferenceBins = alertPostInferenceBinsBySensor[publicDeviceId]

File: backend/anomaly_backend/routes/post_inference_bins.py
from datetime import datetime
from typing import Annotated, cast

from fastapi import APIRouter, Depends, Query
from pydantic import ValidationError
from sqlalchemy.engine import RowMapping
from sqlalchemy.ext.asyncio import AsyncConnection

from anomaly_backend.contracts import (
    CorpusDeviceId,
    HistoricalDateTime,
    PostInferenceBin,
    PostInferenceBinSource,
    PostInferenceBinsQuery,
    PostInferenceBinsResponse,
    format_historical_datetime,
    make_keyset_cursor,
    parse_keyset_cursor,
)
from anomaly_backend.db import get_connection
from anomaly_backend.problems import InvalidQuery, new_request_id
from anomaly_backend.sql import (
    live_post_inference_bin_rows,
    post_inference_bin_rows,
)

router = APIRouter()


def _datetime(row: RowMapping, field: str) -> datetime:
    return cast(datetime, row[field])


def _optional_historical(row: RowMapping, field: str) -> HistoricalDateTime | None:
    value = row[field]
    if value is None:
        return None
    return format_historical_datetime(cast(datetime, value))


@router.get("/api/post-inference-bins", response_model=PostInferenceBinsResponse)
async def post_inference_bins(
    device_id: Annotated[CorpusDeviceId, Query()],
    from_ts: Annotated[HistoricalDateTime, Query(alias="from")],
    to_ts: Annotated[HistoricalDateTime, Query(alias="to")],
    connection: Annotated[AsyncConnection, Depends(get_connection)],
    limit: Annotated[int, Query(ge=1, le=5_000)] = 500,
    cursor: Annotated[str | None, Query()] = None,
    model_version: Annotated[str | None, Query()] = None,
    source: Annotated[PostInferenceBinSource, Query()] = "replay",
) -> PostInferenceBinsResponse:
    query_fields: dict[str, object] = {
        "device_id": device_id,
        "from": from_ts,
        "to": to_ts,
        "source": source,
        "limit": limit,
    }
    if cursor is not None:
        query_fields["cursor"] = cursor
    if model_version is not None:
        query_fields["model_version"] = model_version
    try:
        query = PostInferenceBinsQuery.model_validate(query_fields, strict=True)
    except ValidationError as error:
        raise InvalidQuery("Query parameters failed validation") from error
    filters: dict[str, object] = {
        "device_id": query.device_id,
        "from": query.from_ts,
        "source": query.source,
        "model_version": query.model_version,
    }
    try:
        after = (
            parse_keyset_cursor(
                query.cursor,
                "post_inference_bins",
                snapshot_to=query.to_ts,
                filters=filters,
            )
            if query.cursor
            else None
        )
    except ValueError as error:
        raise InvalidQuery(
            "Query parameters failed validation",
            {"cursor": ["Invalid cursor"]},
        ) from error
    fetch_rows = (
        live_post_inference_bin_rows
        if query.source == "live"
        else post_inference_bin_rows
    )
    canonical_version, rows = await fetch_rows(
        connection,
        device_id=query.device_id,
        from_ts=datetime.fromisoformat(query.from_ts),
        to_ts=datetime.fromisoformat(query.to_ts),
        model_version=query.model_version,
        limit=query.limit,
        after_ts=datetime.fromisoformat(after[0]) if after else None,
        after_id=after[1] if after else None,
    )
    has_more = len(rows) > query.limit
    bins = [
        PostInferenceBin(
            segment_id=cast(int, row["segment_id"]),
            bin_ordinal=cast(int, row["bin_ordinal"]),
            start_score_ts=format_historical_datetime(_datetime(row, "start_score_ts")),
            end_score_ts=format_historical_datetime(_datetime(row, "end_score_ts")),
            scored_timestamp_count=cast(int, row["scored_timestamp_count"]),
            is_alert=cast(bool, row["is_alert"]),
            candidate_alert_count=cast(int, row["candidate_alert_count"]),
            first_alert_ts=_optional_historical(row, "first_alert_ts"),
            last_alert_ts=_optional_historical(row, "last_alert_ts"),
            peak_score=cast(float, row["peak_score"]),
            latest_score=cast(float, row["latest_score"]),
            threshold=cast(float, row["threshold"]),
            schema_version=cast(str, row["schema_version"]),
        )
        for row in rows[: query.limit]
    ]
    return PostInferenceBinsResponse.model_validate(
        {
            "request_id": new_request_id(),
            "device_id": query.device_id,
            "from": query.from_ts,
            "to": query.to_ts,
            "time_zone": "Asia/Jakarta",
            "source": query.source,
            "model_version": canonical_version,
            "bins": bins,
            "next_cursor": (
                make_keyset_cursor(
                    "post_inference_bins",
                    timestamp=format_historical_datetime(
                        _datetime(rows[query.limit - 1], "cursor_ts")
                    ),
                    row_id=cast(str, rows[query.limit - 1]["row_id"]),
                    snapshot_to=query.to_ts,
                    filters=filters,
                )
                if has_more
                else None
            ),
            "returned_count": len(bins),
        },
        strict=True,
    )
